lib/improver/utilities/indexing_operations.py

```diff
@@ -0,0 +1,168 @@
+    The figure below illustrates the results of this intermediate broadcasting
+    step.
+
+    .. figure:: extended_documentation/utilities/indexing_operations/
+       numpy_choose_broadcasting.png
+       :scale: 60 %
+       :alt: Diagram to help explain broadcasting behaviour of choose.
+
+       The form of the index_arrays after they have been broadcast to match the
+       sub-arrays from which values are being taken.
+
+    Args:
+        index_array (np.array of ints):
+            This array must contain integers in the range [0, n-1], where n is
+            the number of arrays in the array_set (equivalent to the length of
+            the leading dimension of array_set if it is a multi-dimensional
+            array rather than a list).
+        array_set (np.array):
+            A multi-dimensional array, where the leading dimension is in effect
+            an indexing dimension. Within this leading dimension are the
+            sub-arrays from which values are to be extracted at positions that
+            match those given in the index_array.
+    Returns:
+        result (np.array):
+            An array containing data extracted from the array_set. The returned
+            array will have the same shape as the sub-arrays unless the
+            index_array has dimensions equivalent to array_set. In this case
+            the returned array will have an equivalent number of dimensions
+            (i.e. one more than the sub-arrays).
+    Raises:
+        ValueError: If input arrays cannot be broadcast to a common shape.
+        IndexError: If dimensions are added to array_set during the
+                    broadcasting of arrays; this will prevent the later
+                    indexing from working as expected.
+        IndexError: If an index exceeds the number of available sub-arrays.
+    """
+    # Used to correct array shape in output following intermediate broadcast.
+    index_array = np.array(index_array)
+    index_dims = index_array.ndim
+    trim_result = index_array.shape != array_set.shape
+
+    # Broadcast arrays to matching shape. This is included to replicate as much
+    # of the numpy choose functionality as possible.
+    try:
+        broadcast_arrays = np.broadcast_arrays(index_array, array_set)
+    except ValueError as err:
+        msg = '{}\nindex_array shape: {}\narray_set shape: {}'.format(
+            err, index_array.shape, array_set.shape)
+        raise ValueError(msg)
+
+    # numpy choose can handle the condition below, but we currently have no
+    # need to accommodate this behaviour, so we trap it and raise an error.
+    if broadcast_arrays[1].ndim > array_set.ndim:
+        msg = ('Dimensionality of array_set has increased which will prevent '
+               'indexing from working as expected.')
+        raise IndexError(msg)
+
+    index_array = broadcast_arrays[0]
+    array_set = broadcast_arrays[1]
+
+    # Use indexing to extract the requires values from the sub-arrays.
```

Comment:
Typo `required`.  Suggest `Use index_array to reorder values from array_set along the required dimension into a new array`.

Reply:
Removed.